### PR TITLE
Upgrade to RAJA v0.14 and use RAJA::inclusive_scan in HiOp

### LINKED ISSUE #342: Upgrade RAJA for CI tests
In order to use RAJA::inclusive_scan, HiOp requires RAJA version 0.14 or greater. 

Some HiOp matrix/vector kernel functions need to be operated on the host. 
Using RAJA::inclusive_scan can help us to move all these computations to the device. 

@ashermancinelli Can you upgrade RAJA to the latest release? It should work with the `develop` branch. I will modify the files where I can use `RAJA::inclusive_scan`, and file another PR.

### PR DESCRIPTION
**Pull request by nychiang**
Upgrade to RAJA v0.14 and use RAJA::inclusive_scan in HiOp

NOTE that this PR must be applied AFTER #322  and #333 

CLOSE #342

### CODE REVIEW

**Comment by nychiang** [comment]
@ashermancinelli @cnpetra @pelesh 
I rebased this branch to the latest `develop`, and it passed all the CI tests.
Since now we are using RAJA v0.14, I think this PR is ready.

File: src/LinAlg/hiopMatrixRajaSparseTriplet.cpp [hunk 1]
```diff
@@ -1648,13 +1716,11 @@ void hiopMatrixRajaSparseTriplet::setSubmatrixToConstantDiag_w_rowpattern(const
   assert(nrm == nnz_to_copy);
 #endif
 
-#if 0 // implemenation that requires a RAJA new version released in Nov.2021
+#if 1 // implemenation that requires a RAJA new version released in Nov.2021
```
Comment (cnpetra) [inline src/LinAlg/hiopMatrixRajaSparseTriplet.cpp:1719]: fix "implemenation" typo (multiple occurrences)

**Review by cnpetra — APPROVED** [review]

File: src/LinAlg/hiopMatrixRajaSparseTriplet.cpp [hunk 2]
```diff
@@ -75,6 +75,13 @@
 #include <cassert>
 // #include <numeric> //std::inclusive_scan is only available after C++17
 
+#define MY_RAJA_GPU_BLOCK_SIZE 128
+
+using my_raja_exec   = RAJA::cuda_exec<MY_RAJA_GPU_BLOCK_SIZE>;
+
+#define RAJA_LAMBDA [=] __device__
```
Comment (ashermancinelli) [inline src/LinAlg/hiopMatrixRajaSparseTriplet.cpp:82]: I don't see this execution policy used anywhere in this PR, so I'm a bit confused about why it's here. The block size, exec policy, and the `RAJA_LAMBDA` macro should already be defined elsewhere (`hiop_raja_defs.hpp`?).

File: src/LinAlg/hiopMatrixRajaSparseTriplet.cpp [hunk 3]
```diff
@@ -1088,8 +1122,9 @@ void hiopMatrixRajaSparseTriplet::copyRowsBlockFrom(const hiopMatrix& src_gen,
     src.row_starts_ = src.allocAndBuildRowStarts();
   }
   assert(src.row_starts_);
-  index_type* src_row_st_host = src.row_starts_->idx_start_host_;
 
+#if 0 // cpu code due to the RAJA::scan bug
```
Comment (ashermancinelli) [inline src/LinAlg/hiopMatrixRajaSparseTriplet.cpp:1126]: We can always look in the git history to find the cpu implementation if we need it, I recommend we remove this ifdef-ed block.

File: src/LinAlg/hiopMatrixRajaSparseTriplet.cpp [hunk 4]
```diff
@@ -1545,13 +1615,11 @@ void hiopMatrixRajaSparseTriplet::setSubmatrixToConstantDiag_w_colpattern(const
   assert(nrm == nnz_to_copy);
 #endif
 
-#if 0 // implemenation that requires a RAJA new version released in Nov.2021
+#if 1 // implemenation that requires a RAJA new version released in Nov.2021
```
Comment (ashermancinelli) [inline src/LinAlg/hiopMatrixRajaSparseTriplet.cpp:1618]: Same comment about this being `ifdef`ed out. If the impls are equivilant and we would like to keep both around, I would prefer to have a macro at the top of this source file to control the behavior, something like `HIOP_MATRIX_RAJA_SP_TRIPLET_USE_RAJA_INCLUSIVE_SCAN` (naming is hard).

**Review by ashermancinelli — CHANGES_REQUESTED** [review]
I would really like either the cpu version removed, or some way to toggle whether the inclusive scan version is used.

**Comment by nychiang** [comment]
@asher After discussed with Cosmin, I deleted the CPU implementations and added some comments about the latest CPU implementation.

**Comment by cnpetra** [comment]
I guess this is ready for merging?

**Comment by nychiang** [comment]
> I guess this is ready for merging?

I thought this had been merged. Let me rebase this PR to the latest develop branch and make sure there is no conflicts.

**Comment by nychiang** [comment]
> I thought this had been merged. Let me rebase this PR to the latest develop branch and make sure there is no conflicts.

Just confirmed that this branch is safe to be merged into the latest `develop`

**Review by cnpetra — APPROVED** [review]

**Pull request merged**